WIP: Apply fast array indexing for Vehicles and Objects

LuXorioN96 opened the pull request:
This Pull Request extends Pool Performance for Objects & Vehicles (previous PR: https://github.com/multitheftauto/mtasa-blue/pull/480)

Regards
Luxy.c

CrosRoad95 commented:
did you do any test with performance before and after?

ghost commented:
No, but i'm currently writing a resource to test performance - will be attached once written.

- LuX

CrosRoad95 commented:
can you say results of both tests?

ghost commented:
Fixed object creation and also fixed some stuff

ghost commented:
done

ghost commented:
> can you say results of both tests?

@CrosRoad95  He can't because it will be seriously flawed as it's not complete. The main performance difference will be seen during gameplay where collisions will not cause lags and all other event handlers in CClientGame. 

**BEFORE:**
```
bool CClientGame::ObjectBreakHandler(CObjectSAInterface* pObjectInterface, CEntitySAInterface* pAttackerInterface)
{
    if (pObjectInterface)
    {
        // Get our object and client object
        CObject*       pObject = g_pGame->GetPools()->GetObjectA((DWORD*)pObjectInterface);
        CClientObject* pClientObject = m_pManager->GetObjectManager()->GetSafe(pObject);

        // Is our client vehicle valid?
        if (pClientObject)
        {
            if (!pClientObject->IsBreakable(false))
                return false;

            // ... //
            // ... //
            // ... //
        }
    }
    return true;
}
```

**AFTER:**
```
SClientEntity* pObjectEntity = g_pGame->GetPools()->GetObjectA((DWORD*)pObjectInterface);
CObjectSA * pObject = pObjectEntity->pEntity;
if (pObject)
{
    CClientObject* pClientObject = pObjectEntity->pClientEntity;
}
```
As you can see we are NOT even looping once, we access the value by index only in improved version.
Currently, MTA will execute 2 separate loops for searching the right object. The more objects you have, the more slow the operation will be.

In the upcoming PR #480, despite of how many objects we have, it will never slow down. It will take the same time to access any object because`g_pGame->GetPools()->GetObjectA` will not loop through array, it will only use the index.

There are many advantages to this but two are very easily noticeable:
- Maps will load much faster
- Improved collision performance 

 I will record a video and post proper benchmarks when it's done.

Einheit-101 commented:
I can easily test this within a few minutes by simply loading the biggest map on my server and then checking the fps with 2 different MTA installations, one with and one without this patch.

ghost commented:
@Einheit-101 If you want, you can check with current nightly - changes for this might be done on sunday or even earlier - who knows 👅 

- LuX

Einheit-101 commented:
I made performance tests with current version, i will repeat them as soon as i have a build with this optimization applied

qaisjp commented:
https://github.com/multitheftauto/mtasa-blue/pull/506/files shows lots of whitespace changes. Please fix this!

codenulls committed:
Fix crashes, bugs, and formatting

Some code was written by lucy.x and some was written by me. I combined them in one commit to fix the formatting.

codenulls committed:
Improve variable names

xLuxy committed:
[NOT WORKING] Changed CPoolsSA::GetPed return type to SClientType<CEntity>*

xLuxy committed:
Fixed m_pedPool causing crash

xLuxy committed:
Fixed m_pedPool completely including event calls - no more crashes :-)

xLuxy committed:
Merge branch 'fast_lookup_byindex' of https://github.com/LuXorioN96/mtasa-blue into fast_lookup_byindex

# Conflicts:
#	Client/game_sa/CPoolsSA.cpp
#	Client/mods/deathmatch/logic/CClientGame.cpp

xLuxy committed:
Merge branch 'fast_lookup_byindex' of https://github.com/LuXorioN96/mtasa-blue into fast_lookup_byindex

# Conflicts:
#	Client/game_sa/CPoolsSA.cpp
#	Client/mods/deathmatch/logic/CClientGame.cpp

xLuxy committed:
Merge branch 'fast_lookup_byindex' of https://github.com/LuXorioN96/mtasa-blue into fast_lookup_byindex

xLuxy committed:
Addendum to previous commit - Fixed merge conflicts

xLuxy committed:
Changed CPoolsSA::GetVehicle return type to SClientType<CEntity>*

xLuxy committed:
Fixed typo in previous commit

xLuxy committed:
* Changed CPoolsSA::GetObject return type to SClientType<CEntity>*
* Fixed crash when starting to aim with a weapon

xLuxy committed:
Fixed crash when jumping on a Vehicle

xLuxy committed:
Some more fixes

xLuxy committed:
Added pointer validations

xLuxy committed:
Fixed incorrect attacker references being passed to lua

xLuxy committed:
Removed CGameEntityXRefManagerImpl class entirely (It crashes when reconnecting - still investigating)
